### ISSUE TITLE
영상 데이터 추출 방법

### ISSUE THREAD

**Issue by SeBin7** [SeBin7/Road_Vision#21]
### ** 1. 가상환경 세팅 필수(기존 가상환경이 편함)**

```
python3 -m venv yt-env

source yt-env/bin/activate

pip install --upgrade pip

pip install yt-dlp

#설치확인
python -c "import yt_dlp; print(yt_dlp.__version__)"

#시스템에 설치
sudo apt install ffmpeg
#설치확인
ffmpeg -version

```

### **2.  유튜브 영상 다운로드용 파이썬 파일 작성**

```
import yt_dlp

url = 'https://www.youtube.com/watch?v=l0U1CaOB9Z4&list=PL26w7arZdurV-lsFxCVL9zd9j9zrPsM4A&index=21'

ydl_opts = {
    'outtmpl': 'downloads/%(title)s.%(ext)s',
    'format': 'bv*[vcodec*=avc1]+ba/best[ext=mp4]/best',  # AV1 제외
    'merge_output_format': 'mp4',
    'postprocessors': [{
        'key': 'FFmpegVideoConvertor',
        'preferedformat': 'mp4',
    }]
}

with yt_dlp.YoutubeDL(ydl_opts) as ydl:
    ydl.download([url])

```

### **3. 실행**

python download_video.py

**Comment by KYJ28** [comment]
best_acc = 0.0
    patience = 3
    counter = 0

    for epoch in range(5):
        cnn.train() 
        #학습 성능 향상
        cnn, optimizer = ipex.optimize(
             model=cnn,
             optimizer=optimizer,
             dtype=torch.float32,
             inplace=True
         )

        classifier.train()
        classifier, optimizer = ipex.optimize(
             model=classifier,
             optimizer=optimizer,
             dtype=torch.float32,
             inplace=True
         )

        total_loss, correct, total = 0, 0, 0
        for batch_idx, batch in enumerate(dataloader):
            if batch is None:
                continue
            windows, labels = batch
            b, win, c, h, w = windows.size()
            windows = windows.view(b * win, c, h, w).to(device)
            labels = labels.to(device)

            features = cnn(windows, flatten=True)
            features = features.view(b, win, -1)
            outputs = classifier(features)

            loss = criterion(outputs, labels)
            optimizer.zero_grad()
            loss.backward()
            optimizer.step()

            total_loss += loss.item() * b
            preds = torch.argmax(outputs, dim=1)
            correct += (preds == labels).sum().item()
            total += b

            print(f"[E{epoch+1}] B{batch_idx}/{len(dataloader)}: Loss={loss.item():.4f}, Acc={correct/total:.4f}")

##early stopping 추가##       
 epoch_acc = correct / total
        epoch_loss = total_loss / total
        print(f"[Epoch {epoch+1}] 평균 Loss: {epoch_loss:.4f}, Acc: {epoch_acc:.4f}\n")

        if epoch_acc > best_acc:
            print("🔼 모델 개선됨: 저장합니다")
            best_acc = epoch_acc
            counter = 0
            torch.save(cnn.state_dict(), "best_cnn_feature_extractor.pth")
            torch.save(classifier.state_dict(), "best_gru_mlp_classifier.pth")
            print("Saved state_dict (.pth)")
            torch.save(cnn, "best_cnn_feature_extractor.pt")
            torch.save(classifier, "best_gru_mlp_classifier.pt")
            print("Saved full model (.pt)")

            # CNN만 ONNX로 export (Hailo-8용)
            cnn.eval()
            dummy_cnn_input = torch.randn(1, 3, 224, 224).to(device)

            torch.onnx.export(
                cnn,
                dummy_cnn_input,
                "cnn_feature_extractor.onnx",
                input_names=["input"],
                output_names=["feature"],
                opset_version=11
            )

            print("✅ CNN 모델을 ONNX 형식으로 저장 완료 (cnn_feature_extractor.onnx)")

        else:
            counter += 1
            print(f"😕 정확도 개선 없음 (카운트 {counter}/{patience})")
            if counter >= patience:
                print("🛑 Early stopping 발동: 학습 즉시 종료")
                return  # 🚨 즉시 전체 학습 종료
